### PR TITLE
Improve and fix `dpkg` algorithm in version compare

### LINKED ISSUE #31483: VD - Revise version matcher logic
# Description 

It was found as part of this issue https://github.com/wazuh/wazuh/issues/31359 some invalid characters message warnings during the version comparison stage in a vulnerability scan. 

```console
2025/08/19 12:07:29 wazuh-modulesd:vulnerability-scanner[72568] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
```

The logic needs to be revised since it should be aligned with dpkg cli. 

```console
root@jammy:/home/vagrant/INDEXER# dpkg --compare-versions 1:4.16.0-2+really2.41-5 lt 1:4.15
root@jammy:/home/vagrant/INDEXER# echo $?
1
root@jammy:/home/vagrant/INDEXER# dpkg --compare-versions 1:4.16.0-2+really2.41-5 gt 1:4.15
root@jammy:/home/vagrant/INDEXER# echo $?
0
```

## DoD 

- [ ] Revise the rational behind the current logic 
- [ ] Align the logic with the CLI results (evaluate if the logic is required and add constraints) 
- [ ] Implement UTs

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
Closes #31483 

## Description

This pull request fixes the **version comparison logic (DPKG)** in the Vulnerability Scanner module.  
Previously, invalid character warnings were logged when comparing versions with revisions containing `+really`, e.g.:

```console
2025/08/19 12:07:29 wazuh-modulesd:vulnerability-scanner[72568] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
````

The logic has been updated to properly align with the **dpkg CLI** behavior, ensuring accurate comparisons and preventing false warnings.

## Proposed Changes

* Revised the rationale and implementation of the `dpkg` logic.
* Ensured alignment with `dpkg --compare-versions` results using C++17.
* Implemented new **unit tests** to validate the corrected behavior.

### Results and Evidence

```console
Input: "1:4.16.0-2+really2.41-5"
- epoch: 1
- version: "4.16.0-2+really2.41"
- revision: "5"
```

* UTs pass with correct dpkg-aligned results.

### Manual tests with their corresponding evidence

* Verified comparisons against `dpkg --compare-versions` for multiple cases.
* Confirmed absence of debug warnings for revisions with `+really`.
* [x] Unit tests executed successfully.

### Artifacts Affected

* Vulnerability Scanner version comparison logic.
* Unit tests related to version parsing.

### Configuration Changes

* N/A

### Tests Introduced

* Unit tests for edge cases in version comparisons (e.g., revisions with `+really`).

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp [hunk 1]
```diff
@@ -45,6 +47,14 @@ class VersionObjectDpkg final : public IVersionObject
...
     std::string m_revision;
 
+    /**
+     * Check if character is blank (space or tab only, like c_isblank)
+     */
+    bool isBlank(char c) const
+    {
+        return c == ' ' || c == '\t';
+    }
+
```
Comment (JcabreraC) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Unused function

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp [hunk 2]
```diff
@@ -16,12 +16,14 @@
 #include "stringHelper.h"
 #include "vulnerabilityScannerDefs.hpp"
 #include <cctype>
+#include <charconv>
```
Comment (JcabreraC) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Is this header necessary? We no longer use `std::from_chars`.

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp [hunk 3]
```diff
@@ -161,116 +171,174 @@ class VersionObjectDpkg final : public IVersionObject
...
+
+        auto is_valid_upstream = [&](std::string_view v) -> bool
+        {
+            if (v.empty())
             {
                 return false;
             }
-            if (!*++colon)
+            // Keep stricter "must start with digit" behavior:
+            if (!std::isdigit(static_cast<unsigned char>(v.front())))
```
Comment (JcabreraC) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Is this restriction correct ?
According to the Debian Policy Manual, upstream versions may start with non-numeric characters.

Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Thanks, you're right! The upstream_version may contain only alphanumerics and the characters . + - : ~ (full stop, plus, hyphen, colon, tilde) and should start with a digit

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp (hunk 2 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Done!

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Done!

**Comment by MiguelazoDS** [comment]
# Testing  :green_circle: 

## Before 

image

```console
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] packageScanner.hpp:315 at platformVerify(): DEBUG: The platform is in the list based on OS code name comparison for Package: login, Version: 1:4.16.0-2+really2.41-5, CVE: CVE-2023-4641, Content OS code name: trixie, OS code name: trixie
```

```console
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
2025/09/02 18:36:04 wazuh-modulesd:vulnerability-scanner[130885] versionObjectDpkg.hpp:265 at match(): DEBUG: Invalid character in revision: 2+really2.41-5 in 1:4.16.0-2+really2.41-5.
``` 

```json
[
  "libpython3.13-minimal",
  "CVE-2025-8194"
]
[
  "libpython3.13-minimal",
  "CVE-2025-6069"
]
[
  "python3.13-minimal",
  "CVE-2025-8194"
]
[
  "python3.13-minimal",
  "CVE-2025-6069"
]
[
  "python3.13",
  "CVE-2025-8194"
]
[
  "python3.13",
  "CVE-2025-6069"
]
[
  "python3.13-venv",
  "CVE-2025-8194"
]
[
  "python3.13-venv",
  "CVE-2025-6069"
]
[
  "libpython3.13",
  "CVE-2025-8194"
]
[
  "libpython3.13",
  "CVE-2025-6069"
]
[
  "libpython3.13-dev",
  "CVE-2025-8194"
]
[
  "libpython3.13-dev",
  "CVE-2025-6069"
]
[
  "python3.13-dev",
  "CVE-2025-8194"
]
[
  "python3.13-dev",
  "CVE-2025-6069"
]
[
  "libpython3.13-stdlib",
  "CVE-2025-8194"
]
[
  "libpython3.13-stdlib",
  "CVE-2025-6069"
]
[
  "axios",
  "CVE-2021-3749"
]
[
  "axios",
  "CVE-2020-28168"
]
[
  "axios",
  "CVE-2025-27152"
]
[
  "axios",
  "CVE-2019-10742"
]
```

## After 

Package is now scanned 
```console
2025/09/02 18:49:34 wazuh-modulesd:vulnerability-scanner[150839] packageScanner.hpp:412 at versionMatch(): DEBUG: Scanning package - 'login' (Installed Version: 1:4.16.0-2+really2.41-5, Security Vulnerability: CVE-2004-1001). Identified vulnerability: Version: 0. Required Version Threshold: 1:4.0.3-35. Required Version Threshold (or Equal): .
2025/09/02 18:49:34 wazuh-modulesd:vulnerability-scanner[150839] packageScanner.hpp:562 at versionMatch(): DEBUG: No match due to default status for Package: login, Version: 1:4.16.0-2+really2.41-5 while scanning for Vulnerability: CVE-2004-1001
``` 

No warnings or errors 

image

```json 
[
  "libpython3.13-minimal",
  "CVE-2025-8194"
]
[
  "libpython3.13-minimal",
  "CVE-2025-6069"
]
[
  "python3.13-minimal",
  "CVE-2025-8194"
]
[
  "python3.13-minimal",
  "CVE-2025-6069"
]
[
  "python3.13",
  "CVE-2025-8194"
]
[
  "python3.13",
  "CVE-2025-6069"
]
[
  "python3.13-venv",
  "CVE-2025-8194"
]
[
  "python3.13-venv",
  "CVE-2025-6069"
]
[
  "libpython3.13",
  "CVE-2025-8194"
]
[
  "libpython3.13",
  "CVE-2025-6069"
]
[
  "libpython3.13-dev",
  "CVE-2025-8194"
]
[
  "libpython3.13-dev",
  "CVE-2025-6069"
]
[
  "python3.13-dev",
  "CVE-2025-8194"
]
[
  "python3.13-dev",
  "CVE-2025-6069"
]
[
  "libpython3.13-stdlib",
  "CVE-2025-8194"
]
[
  "libpython3.13-stdlib",
  "CVE-2025-6069"
]
[
  "axios",
  "CVE-2021-3749"
]
[
  "axios",
  "CVE-2020-28168"
]
[
  "axios",
  "CVE-2025-27152"
]
[
  "axios",
  "CVE-2019-10742"
]
```

File: src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp [hunk 4]
```diff
@@ -615,6 +612,183 @@ TEST_F(VersionMatcherTest, compareSemVer_OkGreater)
     EXPECT_EQ(compareResult, VersionComparisonResult::A_GREATER_THAN_B);
 }
 
+TEST_F(VersionMatcherTest, aditional_dpkgScenarios)
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp]: Additional

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp [hunk 5]
```diff
@@ -161,116 +161,174 @@ class VersionObjectDpkg final : public IVersionObject
...
+     * @param out Dpkg object to store the result.
      * @return true/false according to match condition.
      */
-    static bool match(const std::string& version, Dpkg& output)
+    static bool match(const std::string& input, Dpkg& out)
     {
-        const char *end, *ptr;
-        const char* string = version.c_str();
+        // Use string_view for efficient string operations (available in C++17)
+        const std::string_view ASCII_WS = " \t\r\n\f\v";
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Trying to follow the changes. Before we were only testing trailing and leading space. Do we found these extra cases?

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:1]: Aren't we losing information here after the removal of the "Invalid character..." log messages?

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp (hunk 5 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp]: Thanks, improve it!

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:1]: Thanks added !

File: src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp (hunk 4 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp]: Thanks!

**Review by MiguelazoDS — APPROVED** [review]

**Comment by MiguelazoDS** [comment]
## Test :green_circle: 

image

```json
[
  "libpython3.13-minimal",
  "CVE-2025-8194"
]
[
  "libpython3.13-minimal",
  "CVE-2025-6069"
]
[
  "python3.13-minimal",
  "CVE-2025-8194"
]
[
  "python3.13-minimal",
  "CVE-2025-6069"
]
[
  "python3.13",
  "CVE-2025-8194"
]
[
  "python3.13",
  "CVE-2025-6069"
]
[
  "python3.13-venv",
  "CVE-2025-8194"
]
[
  "python3.13-venv",
  "CVE-2025-6069"
]
[
  "libpython3.13",
  "CVE-2025-8194"
]
[
  "libpython3.13",
  "CVE-2025-6069"
]
[
  "libpython3.13-dev",
  "CVE-2025-8194"
]
[
  "libpython3.13-dev",
  "CVE-2025-6069"
]
[
  "python3.13-dev",
  "CVE-2025-8194"
]
[
  "python3.13-dev",
  "CVE-2025-6069"
]
[
  "libpython3.13-stdlib",
  "CVE-2025-8194"
]
[
  "libpython3.13-stdlib",
  "CVE-2025-6069"
]
[
  "axios",
  "CVE-2021-3749"
]
[
  "axios",
  "CVE-2020-28168"
]
[
  "axios",
  "CVE-2025-27152"
]
[
  "axios",
  "CVE-2019-10742"
]
``` 

Package scanned 
image

```console
2025/09/04 15:44:45 wazuh-modulesd:vulnerability-scanner[1044658] packageScanner.hpp:412 at versionMatch(): DEBUG: Scanning package - 'login' (Installed Version: 1:4.16.0-2+really2.41-5, Security Vulnerability: CVE-2004-1001). Identified vulnerability: Version: 0. Required Version Threshold: 1:4.0.3-35. Required Version Threshold (or Equal): .
2025/09/04 15:44:45 wazuh-modulesd:vulnerability-scanner[1044658] packageScanner.hpp:562 at versionMatch(): DEBUG: No match due to default status for Package: login, Version: 1:4.16.0-2+really2.41-5 while scanning for Vulnerability: CVE-2004-1001
```

**Review by MiguelazoDS — APPROVED** [review]
So much clear now!, I was thinking in the "+" sign at first, but it's the duplicate hyphen. Good  catch

**Comment by JcabreraC** [comment]
With this, the issue is fixed. Example:

### Before
In the previous version of the code, parsing the version `1:4.16.0-2+really2.41-5` was incorrect:
- **Epoch:** `1` ✅ (correct)
- **Upstream:** `4.16.0-2+really2.41` ❌ (incorrect)
- **Revision:** `5` ❌ (incorrect)

The problem was that the code looked for the **first hyphen** to separate upstream and revision, but it should look for the **last hyphen**.


### After
With the fix. Current parsing of `1:4.16.0-2+really2.41-5` is:

- **Epoch:** `1` ✅ (correct)
- **Upstream:** `4.16.0-2+really2.41` ✅ (correct – includes the first hyphen)
- **Revision:** `5` ✅ (correct)

Thus, this PR fixes the original error. 

However, a number of potential improvements have been identified.

## Future Improvements

### 1. Incorrect validation of characters in the version number
```cpp
if (!std::isdigit(*ptr) && !std::isalpha(*ptr) && strchr(".-+:~", *ptr) == NULL)
```
The colon (`:`) is allowed in the upstream version number, but according to the dpkg specification, `:` should only be used as the epoch separator. Once the epoch is parsed, no further colons should appear in the version part.

### 2. Incomplete set of allowed characters in the revision
```cpp
if (!std::isdigit(*ptr) && !std::isalpha(*ptr) && strchr(".+~", *ptr) == NULL)
```
The hyphen (`-`) is not allowed in the revision number, but dpkg does allow it. This can cause false positives in versions like `1.0-ubuntu1-security`.

### 3. Overly restrictive validation at the start of the version
```cpp
if (!std::isdigit(*ptr++))
{
    logDebug1(WM_VULNSCAN_LOGTAG, "Version number does not start with digit");
    return false;
}
```
Debian policy states that versions *should start* with a digit, not *must start*. dpkg accepts versions starting with letters (with a warning). This validation rejects valid versions like `abc1.0-1`.

Reference: [Debian Policy – Version](https://www.debian.org/doc/debian-policy/ch-controlfields.html?utm_source=chatgpt.com#version)

**Review by JcabreraC — APPROVED** [review]
LGTM !

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp
Comment (Dwordcito) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:1]: Remove all debug logs, because the logging helper introduce a lot of performance degradation, memory fragmentation, etc.

Because in loggerHelper.h we convert const char* msg (3th parameter)  to std::string (allocate memory if the size is > to the SSO (small string optimization (16~21 bytes)) during the call to GLOBAL_LOG_FUNCTION.

@GabrielEValenzuela Create an issue for this. I did fix this in 5.0 https://github.com/wazuh/wazuh/commit/7db9c79cc8ff43c1d282f2c37db36672cdce5cb6

File: src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp [hunk 6]
```diff
@@ -118,10 +118,7 @@ TEST_F(VersionMatcherTest, pkgStrategy)
     EXPECT_NO_THROW((compareResult = VersionMatcher::compare("13.3.1", "13.5", VersionMatcherStrategy::PKG)));
     EXPECT_EQ(compareResult, VersionComparisonResult::A_LESS_THAN_B);
 
-    EXPECT_NO_THROW(
-        (compareResult = VersionMatcher::compare("2.0.4_419.3", "2.0.4_419.3", VersionMatcherStrategy::PKG)));
```
Comment (Dwordcito) [inline src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp:122]: What is the reason to delete this test cases ?

Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/tests/unit/versionMatcher_test.cpp:122]: These are invalid `dpkg` versions.

```bash
dpkg --compare-versions 2.0.4_419.3 eq 2.0.4_419.3
dpkg: warning: version '2.0.4_419.3' has bad syntax: invalid character in version number
dpkg: warning: version '2.0.4_419.3' has bad syntax: invalid character in version number
```

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:1]: OMW

Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/versionMatcher/versionObjectDpkg.hpp:1]: Avoid unnecessary `std::string` allocation in `loggerHelper.h` [#31914](https://github.com/wazuh/wazuh/issues/31914)
